[Chart] Add BarChartView and GroupBarChartView

jkpu opened the pull request:
### Description of Change ###
Add BarChartView and GroupBarChartView
Add CircularChartGallery for chart test

### Bugs Fixed ###
None


### API Changes ###
Added:
 - Add BarChartView
 - Add GroupBarChartView

### Behavioral Changes ###
BarChartView screenshot is following images.

![wc-1024-1-2020-04-01-132856](https://user-images.githubusercontent.com/31116585/81247647-82d77680-9055-11ea-8278-8aba72284c36.png)

![wc-1024-1-2020-04-01-132935](https://user-images.githubusercontent.com/31116585/81247660-879c2a80-9055-11ea-8842-f324cafc8d95.png)

![Default_ReferenceLineData](https://user-images.githubusercontent.com/31116585/81247693-a13d7200-9055-11ea-9770-6bf3e51df014.png)


![Horizontal_ReferenceLineData](https://user-images.githubusercontent.com/31116585/81247683-97b40a00-9055-11ea-8c18-5a0dfa5a8c36.png)

![GroupBarChart](https://user-images.githubusercontent.com/31116585/81247720-b2867e80-9055-11ea-9a47-d4f4f338e2e3.png)

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms/BarChartView.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Gets or sets type of bar chart.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public BarChartType BarChartType
```
I think, it should be a member of `AxisOption` and Name is not proper as I understand, I think it named `MajorAxis`

jkpu commented:
> Just a quick scan of it, but that there are more public API provided in this PR. (not only BarChartView and GroupBarChartView) Plus, it may cause a name resolving conflicts because some class names are the same as Xamarin.Forms API. (e.g. `Entry` , `Label`)
> 
> Anyway, great job!

I Changed class name 
Labe -> TextItem
Entry -> DataItem
and changed DataSet and DataSetArray name
DataSet -> DataItemGroup
DataSetArray -> Data

data structure is following 
DataItem -> DataItemGroup -> Data

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/BarChartView.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Gets or sets the visibility of bar background color.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public bool BarBackgroundColorIsVisible
```
BarBackgroundColor property name  already exist in BarDataItemGroup and BarDataItem. This property mean visibility of whole bar background of BarChartView

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/BarChartView.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Gets or sets a position of data value label.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public ValueLabelPosition ValueLabelPosition
```
If ValueLabelPosition move to Entry(DataItem) class, it is hard to use . user should set value position for each Entry(DataItem).  I think almost chart has same value position in a Chart.

jkpu replied:
Done. move to DataItem class

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+        {
+            if (_canvas != null)
+            {
+                DrawContent(_canvas);
```
How about use `Control.Invalidate()`?
``` c#
   Control?.Invalidate();
```

jkpu replied:
Done

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+            }
+        }
+
+        private SKRect[] MeasureCategoryLabels()
```
How about returing 'IEnumerable<SKRect>'? no need to convert to array
```suggestion
        private IEnumerable<SKRect> MeasureCategoryLabels()
```

jkpu replied:
Done. changed all arrary code in Renderer to 'IEnumerable'.

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/CanvasExtension.cs:
```diff
@@ -0,0 +1,95 @@
+                paint.TextSize = (float)(textItem.FontSize * FontCovertConst);
+                paint.IsAntialias = true;
+                var color = textItem.TextColor == Color.Default ? Color.White : textItem.TextColor;
+                paint.Color = SkiaSharp.Views.Forms.Extensions.ToSKColor(color);
```
`ToSKColor` is extension method
```suggestion
                paint.Color = color.ToSKColor();
```

jkpu replied:
Done

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                            Style = SKPaintStyle.Fill,
+                            Color = SkiaSharp.Views.Forms.Extensions.ToSKColor(barColor),
+                        })
+                        {
```
Why not use DrawLine with `StrokeWidth = barSize.Width`?

jkpu replied:
Is it possible to draw rounded line using DrawLine. I only used DrawRect and DrawRoundRect for drawing radius bar.

myroot replied:
Yes, you can draw rounded line with SKStrokeCap.Round

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                            if (bottomRadius > 0)
+                            {
+                                x = _majorAxisSize.Width;
+                                w = +bottomRadius * 2;
```
```suggestion
                                w = bottomRadius * 2;
```

jkpu replied:
Done

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+            for (int i = 0; i < items.Count; i++)
+            {
+                var index = items[i].Key == 0 ? i : items[i].Key - 1;
+                index = index < _categoryCount ? index : _categoryCount - 1;
```
```suggestion
                index = Math.Min(index, _categoryCount - 1);
```

jkpu replied:
Done

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+            {
+                var index = items[i].Key == 0 ? i : items[i].Key - 1;
+                index = index < _categoryCount ? index : _categoryCount - 1;
+                _barChartDataTable[index] = items[i] as DataItem;
```
Why do you cast as DataItem instead of IDataItem? It is not working if User implement a owned class that is implementing IDataItem
```suggestion
                _barChartDataTable[index] = items[i] as IDataItem;
```

jkpu replied:
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/DataItem.cs:
```diff
@@ -0,0 +1,63 @@
+
+        /// <summary>
+        /// Gets or sets a key.
+        /// Key represents the position of a chart. if the key is 3, this value set 3rd data in a chart.
```
Key means position of data. so it should be 1 instead 0.

myroot replied:
data position could be started at 0, and it is natural on programming language

jkpu replied:
removed Key in DataItem.

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/ChartView.cs:
```diff
@@ -0,0 +1,108 @@
+        }
+
+        /// <summary>
+        /// Gets or sets a value range of data set.
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/AxisOption.cs:
```diff
@@ -0,0 +1,182 @@
+        /// <summary>
+        /// Constructor a new AxisOption structure
+        /// </summary>
+        public AxisOption(bool uniformValue) : this(uniformValue, uniformValue, uniformValue, uniformValue, uniformValue)
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/AxisOption.cs:
```diff
@@ -0,0 +1,182 @@
+        /// For example, If the chart represents a change in the value of a week, each category becomes the day of the week.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public bool IsVisibleOfCategoryLabel { get; set; }
```
Done. removed this property

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/AxisOption.cs:
```diff
@@ -0,0 +1,182 @@
+        /// Gets or sets a list of reference datas.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public IList<Entry> ReferenceDatas { get; set; }
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/Entry.cs:
```diff
@@ -0,0 +1,63 @@
+        /// If Entry.Color is not set, the color is set DataSet.Color.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public Color Color { get; set; } = Color.Default;
```
As we discussed before. I keep current status

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/BarChartView.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Gets or sets the visibility of value label.
+        /// </summary>
+        /// <since_tizen> 4 </since_tizen>
+        public bool ValueLabelIsVisible
```
Done. removed this property

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms/GroupBarChartView.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public static readonly BindableProperty GroupBarMarginProperty = BindableProperty.Create(nameof(GroupBarMargin), typeof(double), typeof(GroupBarChartView), 5d);
+
+        public int GroupDataCount
```
Done. remove GroupDataCount property

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+            if (Element.AxisOption.IsVisibleOfCategoryLabel)
+            {
+                _categoryLabelCount = Element.AxisOption.CategoryLabels.Count();
+                _categoryCount = _categoryLabelCount > _dataCount ? _categoryLabelCount : _dataCount;
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                return;
+            }
+
+            if (Element.AxisOption != null)
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+
+            if (Element.AxisOption.IsVisibleOfCategoryLabel)
+            {
+                _categoryLabelCount = Element.AxisOption.CategoryLabels.Count();
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+            var items = Element.Data[0].DataItems;
+            for (int i = 0; i < items.Count; i++)
+            {
+                var index = items[i].Key == 0 ? i : items[i].Key - 1;
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+        protected virtual SKPoint[] CalculatePoints(SKSize barSize)
+        {
+            var result = new List<SKPoint>();
+            bool isVertical = Element.BarChartType == BarChartType.Vertical ? true : false;
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                {
+                    var x = yAxisWidth + _barHmargin + (barSize.Width / 2) + (barSize.Width + _barHmargin) * i;
+                    double value = _barChartDataTable[i] != null ? Math.Abs(_barChartDataTable[i].Value) : 0;
+                    var y = value >= Element.Maximum ? 0 : (float)(((Element.Maximum - value) / Element.ValueRange) * barSize.Height);
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                {
+                    double value = _barChartDataTable[i] != null ? Math.Abs(_barChartDataTable[i].Value) : 0;
+                    var y =  _barVmargin + (barSize.Width / 2) + (barSize.Width + _barVmargin) * i;
+                    var x = value >= Element.Maximum ? yAxisWidth + barSize.Height : yAxisWidth + (float)(((value - Element.Minimum) / Element.ValueRange) * barSize.Height);
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+
+        const float TextHorizontalMargin = 5;
+        const float TextVerticalMargin = 8;
+        const float FontCovertConst = (float)3;
```
Done. changed from FontCovertConst to Forms.ConvertToEflFontPoint

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,741 @@
+                return;
+            }
+
+            if (Element.AxisOption != null)
```
Not yet fixed

jkpu replied:
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,741 @@
+
+        protected virtual void GenerateDataTable()
+        {
+            if (Element.Data == _prevDataSetArray)
```
Done,  remove _prevDataSetArray  checking code because it can't check the change of  dataitem.

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,752 @@
+                    paint.TextSize = (float)(e.Label.FontSize * FontCovertConst);
+                    paint.MeasureText(text, ref bounds);
+                    return bounds;
+                }).ToArray();
```
When I removed ToArray(), It makes crash when entering test app. So I remain this code.

myroot replied:
Could you tell me which exception was thrown?

jkpu replied:
CallStack  doesn't provide in sdb dlog. and also I can't extract log file from crash dump file because file always is damaged with this crash. you can easily reproduce this issue with CircularUIChartGallery app with removing ToArray().

jkpu replied:
Done. moved using( var paint .. ) code within select code.

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/GroupBarChartViewRenderer.cs:
```diff
@@ -0,0 +1,406 @@
+                var items = Element.Data.DataItemGroups[i].DataItems;
+                for (int j = 0; j < items.Count; j++)
+                {
+                    _groupBarChartDataTable[i, j] = items[j]?.Value ?? Element.Minimum;
```
if items.Count was smaller than _categoryCount, 
`_groupBarChartDataTable[X,N]` not assigned as Element.Minimum value when N was larger than items.Count

jkpu replied:
Done

myroot commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/GroupBarChartViewRenderer.cs:
```diff
@@ -0,0 +1,406 @@
+            for (int i = 0; i < _dataSetArrayCount; i++)
+            {
+                var dataSet = Element.Data.DataItemGroups[i];
+                int entryCount = dataSet.DataItems.Count();
```
```suggestion
                int entryCount = dataSet.DataItems.Count;
```

jkpu replied:
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/GroupBarChartViewRenderer.cs:
```diff
@@ -0,0 +1,406 @@
+                return;
+            }
+
+            if (Element.AxisOption != null)
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,778 @@
+            return new SKSize(barWidth, barHeight);
+        }
+
+        protected virtual IEnumerable<SKPoint> CalculatePoints(SKSize barSize)
```
Done

myroot replied:
I means, Need to update all `IEnumerable<SKPoint> points` to `IList<SKPoint> points`

jkpu replied:
Done. Changed all points IEnumerable<SKPoint> points  -> IList<SKPoint> points

myroot replied:
Now, you can use index accesor(`[i]`) instead `ElementAt(i);`


'points.ElementAt(i); => points[i]'

jkpu replied:
I Changed all related code from  `ElementAt(i)` to index accessor(`[i]`)

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,778 @@
+            {
+                var categoryLabel = Element.AxisOption.CategoryLabels.ElementAt(i);
+                var label = categoryLabel.Label;
+                var index = categoryLabel.ItemIndex != -1 ? categoryLabel.ItemIndex : i ;
```
Done

jkpu commented on src/Tizen.Wearable.CircularUI.Chart.Forms.Renderer/BarChartViewRenderer.cs:
```diff
@@ -0,0 +1,778 @@
+
+            if(barWidth != (float)XForms.ConvertToScaledPixel(Element.BarWidth))
+            {
+                Element.BarWidth = XForms.ConvertToScaledDP((double)barWidth);
```
remove this code and default value of BarWidth  change from 10 to -1. If  user doesn't set BarWidth , renderer caculate BarWidth  using canvas size and category count.
